lib/view_component/compiler.rb

```diff
@@ -103,266 +50,273 @@ def renders_template_for?(variant, format)
-              "found for #{"variant".pluralize(count)} " \
-              "#{duplicate_template_file_and_inline_variant_calls.map { |v| "'#{v}'" }.to_sentence} " \
-              "in #{component_class}. " \
-              "There can only be a template file or inline render method per variant."
-          end
+        errors <<
+          "Template #{"file".pluralize(count)} and inline render #{"method".pluralize(count)} " \
+          "found for #{"variant".pluralize(count)} " \
+          "#{duplicate_template_file_and_inline_call_variants.map { |v| "'#{v}'" }.to_sentence} " \
+          "in #{@component}. There can only be a template file or inline render method per variant."
+      end
 
-          uniq_variants = variants.compact.uniq
-          normalized_variants = uniq_variants.map { |variant| normalized_variant_name(variant) }
+      variant_pairs =
+        @templates.select { _1.variant.present? }.map { [_1.variant, _1.normalized_variant_name] }.uniq { _1.first }
 
-          colliding_variants = uniq_variants.select do |variant|
-            normalized_variants.count(normalized_variant_name(variant)) > 1
-          end
+      colliding_normalized_variants =
+        variant_pairs.map(&:last).tally.select { |_, count| count > 1 }.keys
+          .map do |normalized_variant_name|
+          variant_pairs
+            .select { |variant_pair| variant_pair.last == normalized_variant_name }
+            .map { |variant_pair| variant_pair.first }
+        end
 
-          unless colliding_variants.empty?
-            errors <<
-              "Colliding templates #{colliding_variants.sort.map { |v| "'#{v}'" }.to_sentence} " \
-              "found in #{component_class}."
-          end
+      colliding_normalized_variants.each do |variants|
+        errors << "Colliding templates #{variants.sort.map { |v| "'#{v}'" }.to_sentence} found in #{@component}."
+      end
 
-          errors
-        end
+      raise TemplateError.new(errors) if errors.any? && raise_errors
+
+      errors
     end
 
-    def templates
+    def gather_templates
       @templates ||=
         begin
-          extensions = ActionView::Template.template_handler_extensions
-
-          component_class.sidecar_files(extensions).each_with_object([]) do |path, memo|
+          templates = @component.sidecar_files(
+            ActionView::Template.template_handler_extensions
+          ).map do |path|
             pieces = File.basename(path).split(".")
-            memo << {
+
+            out = Template.new(
+              component: @component,
+              type: :file,
               path: path,
-              format: pieces[1..-2].join(".").split("+").first&.to_sym,
-              variant: pieces[1..-2].join(".").split("+").second&.to_sym,
-              handler: pieces.last
-            }
+              lineno: 0,
+              extension: pieces.last,
+              this_format: pieces[1..-2].join(".").split("+").first&.to_sym,
```

Comment:
we use pieces[1..-2] a few times, which is a bit obtuse in terms of what it represents without extra thought. Thoughts on extracting this (and maybe pieces.last) into a well named intermediate variable?

Reply:
I ended up extracting a different approach here and added some comments. Let me know your thoughts ❤️